server/mvcc/kvstore.go

```diff
@@ -227,7 +227,27 @@ func (s *store) updateCompactRev(rev int64) (<-chan struct{}, int64, error) {
 	return nil, compactMainRev, nil
 }
 
-func (s *store) compact(trace *traceutil.Trace, rev, prevCompactRev int64) (<-chan struct{}, error) {
+// checkPrevCompactionCompleted checks whether the previous scheduled compaction is completed.
+func (s *store) checkPrevCompactionCompleted() bool {
+	tx := s.b.ReadTx()
+	tx.Lock()
```

Comment:
Just curious why not use `tx.RLock()` but understand the PR is a backport of https://github.com/etcd-io/etcd/pull/15985

Reply:
`Lock()` for `ReadTx()` should be by default a `Rlock()`.

Reply:
For my understanding, do you mean the following RWMutex `Lock` is equal to `RLock` to protect the read buffer? 

https://github.com/etcd-io/etcd/blob/5e40a8b00cc1be3ce0dd3d7e6947d417253fb03f/server/storage/backend/read_tx.go#L127-L130

Reply:
After investigating a little bit, I think you are right, we should use RLock() instead. Lock() should only be used when the read buffer is updated by the writes committed by the BatchTxs. However, there are several positions where use read-only operations but Lock() is used. I think it is good to open a separate PR to correct them all.

Reply:
Yep. That aligns with my understanding. 

I also don't think it blocks this PR to be merged. This can be corrected separately.

Reply:
Nice catch @chaochn47 - suggest we raise an issue for this so we don't forget to come back to it.

Reply:
The `Lock()/Unlock()` should only be used in the following three cases,
- When [committing a transaction](https://github.com/etcd-io/etcd/blob/dee90e19f1c6fc2aceed38cd2855010df0b6c6a7/server/storage/backend/batch_tx.go#L336). Because it will reset the read tx, including the buffer;
- When [writing the pending data (not committed to bbolt yet) back to read buf](https://github.com/etcd-io/etcd/blob/dee90e19f1c6fc2aceed38cd2855010df0b6c6a7/server/storage/backend/batch_tx.go#L310), because obviously no reading is allowed when writing the read buffer;
- When [performing defragmentation](https://github.com/etcd-io/etcd/blob/dee90e19f1c6fc2aceed38cd2855010df0b6c6a7/server/storage/backend/backend.go#L464),  because again it will reset the read tx, including the read buffer.

`RLock()/RUnlock()` should be used for all other cases.

Please raise a separate ticket to track the change. It should also be backported to 3.5 and 3.4.

Reply:
@ahrtr thanks for figuring out proper usage of transaction lock. However, I don't_ see reason why we should backport it. Old implementation uses stricter lock, which is sub-optimal, but is not a bug. 

To ensure release stability, we need to assume that the list you proposed can have a mistake. I would prefer that we don't verify correctness of performance improvements on stable releases.

Reply:
Note I prefer to backporting it, but I am not insist on backporting it if there is strong objection from other maintainers.

Reply:
> stricter lock is true, it's also wrong lock. It's a bug to me.

Without a change system behaves correctly, after the change only thing you expect is performance improvement. How this is a bug?

Reply:
I would say it's harmless (on functionality) bug, obviously the lock isn't correctly be used. Let alone it has impact on performance.

Reply:
Bugs is when locks don't protect the critical section, relaxing locks is performance improvement. Nothing is harmless until proven, let alone concurrency change.